src/coreclr/jit/valuenum.cpp

```diff
@@ -9664,6 +9687,25 @@ ValueNumPair ValueNumStore::VNPairForCast(ValueNumPair srcVNPair,
     return {castLibVN, castConVN};
 }
 
+// Compute the ValueNumberPair for an implicit cast on a normalize-on-load assignment
+ValueNumPair ValueNumStore::VNPairForImplicitCastNormalizeOnLoad(ValueNumPair srcVNPair,
+                                                                 var_types    castToType,
+                                                                 var_types    castFromType,
+                                                                 bool         srcIsUnsigned,    /* = false */
+                                                                 bool         hasOverflowCheck) /* = false */
+{
+    ValueNumPair vnPair = VNPairForCast(srcVNPair, castToType, castFromType, srcIsUnsigned, hasOverflowCheck);
+
+    // srcLibVN cannot be the same as castLibVN for normalize-on-load assignments.
+    // So set the castLibVN to be the same as castConVN.
+    if (srcVNPair.GetLiberal() == vnPair.GetLiberal())
+    {
+        vnPair.SetLiberal(vnPair.GetConservative());
+    }
```

Comment:
I do not understand the reasoning here. Conservative/liberal split has to do with memory operations, not casts.

Reply:
I posted a picture of what the diff looks like. I probably didn't describe it correctly. Effectively, I want to use the conservative VN since it is pessimistic.

Reply:
It is "pessimistic" in very particular ways. Conservative VNs reflect the fact memory can change due to other threads mutating it, while liberal ones assume they don't.

So the usage of conservative VNs here does not make sense to me here, since our bug is not caused by optimistically assuming races do not occur.

Reply:
So the only reason convservative VNs exist is for assuming race conditions?

Reply:
Correct.

Reply:
What would be the proper fix for this then?

Reply:
It is a good question to which the answer is not obvious. See also [this gist](https://gist.github.com/SingleAccretion/e10837771d46089a94633668f94344ae) if you haven't read it already.

Essentially, we don't know, in VN, whether the store will _actually_ implicitly truncate. It means we don't know the state of the upper bits. A faithful representation for a lack of knowledge like this is some kind of "new, unique" VN. It will, obviously, run into problems with huge regressions.

So we have to do something smarter, which accounts for the fact we don't know what the upper bits are. For example, one can picture a VNFunc like `VNF_IdontHaveAgoodNameForThis(lower bits VN goes here)`, which would also be "unique", but then on numbering casts we could look through it and see what the lower bits are up to. This may (will) still run into into issues with optimizations which look at SSA descriptors directly (i. e. copy propagation).

Reply:
My idea was to stop VN'ing normalize-on-load locals unless they appear with the cast on top. It means that in your example:
```scala
***** BB04
STMT00003 ( 0x059[E-] ... 0x05D )
N005 (  4,  4)              [000014] -A------R--                         *  ASG       short  $VN.Void
N004 (  2,  2)              [000013] D------N---                         +--*  LCL_VAR   short  V00 arg0         d:4 $VN.Void
N003 (  4,  4) CSE #01 (def)[000011] -----------                         \--*  ADD       int    <l:$41, c:$3c5>
N001 (  2,  2)              [000009] -----------                            +--*  LCL_VAR   int    V00 arg0         u:3 (last use) <l:$46, c:$3c4>
N002 (  1,  1)              [000010] -----------                            \--*  CNS_INT   int    1 $43

***** BB04
STMT00004 ( 0x05F[E-] ... 0x060 )
N003 (  4,  5)              [000016] -----------                         *  RETURN    int    $VN.Void
N002 (  3,  4) CSE #01 (use)[000172] -----------                         \--*  CAST      int <- short <- int <l:$41, c:$3c7>
N001 (  2,  2)              [000015] -----------                            \--*  LCL_VAR   int    V00 arg0         u:4 (last use) <l:$41, c:$3c6>
```
The VN on `[000009]` would be a unique VN and no CSE would be possible.

We would store the full VN in the SSA def but not give it to uses of normalize-onload locals. When we saw the cast on top we would VN it as Cast(stored VN from SSA def).  Alternatively we could store the casted VN in the SSA def, I guess it does not make much of a difference.

It would be expensive because we are going to be assigning a unique VN to all uses of normalize-on-load locals, so some optimization would probably be necessary to make these unique VNs expensive. @SingleAccretion had some ideas here.

Reply:
> had some ideas here.

The starting point was that today all unique VNs take memory for storing loop numbers, so we could introduce some special "phantom" chunks (like those that used to be used for `NotAField` field sequences) for (at least) the `NOT_IN_LOOP` case.

Reply:
This gist https://gist.github.com/SingleAccretion/e10837771d46089a94633668f94344ae - is this still a problem?

Reply:
I believe so. It is the same problem we have been discussing, after all. It is possible the reproduction I have has stopped working since then, though.

Reply:
Trying to understand the problem you all are describing.

Basically, a 'def' `ASG{def}(VAR0{short}, ...)` and a 'use' `CAST(VAR0{int})` - `VAR0` is the same variable with the same liberal value in both the definition and use, but because they are using different sizes, it's technically not correct. This happens today, but the `CAST` is what ensures it does the right thing, which is why removing the `CAST` might be illegal. 

Is this almost right?

Reply:
Yes.

There is another complication that comes with this and it is the fact VN (and IR) do not have true small types. So there is no way to represent a "true" `short` value, only `short`-sign-extended-to-`int`, which means that the "def" values (i. e. those we keep in SSA descriptors) cannot be truly faithful either (this is the "we do not know whether the store truncates or not problem).